serde/src/ser/impls.rs

```diff
@@ -769,20 +769,45 @@ impl Serialize for SystemTime {
 
 ////////////////////////////////////////////////////////////////////////////////
 
+struct Wrapper<'a> {
+    pub buf: &'a mut [u8],
+    pub offset: usize,
+}
+
+impl<'a> fmt::Write for Wrapper<'a> {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        if self.offset > self.buf.len() {
+            return Err(fmt::Error);
+        }
+        let remaining_buf = &mut self.buf[self.offset..];
+        let raw_s = s.as_bytes();
+        let write_num = core::cmp::min(raw_s.len(), remaining_buf.len());
+        remaining_buf[..write_num].copy_from_slice(&raw_s[..write_num]);
+        self.offset += raw_s.len();
+        if write_num < raw_s.len() {
+            Err(fmt::Error)
+        } else {
+            Ok(())
+        }
+    }
+}
+
 /// Serialize a value that implements `Display` as a string, when that string is
 /// statically known to never have more than a constant `MAX_LEN` bytes.
 ///
 /// Panics if the `Display` impl tries to write more than `MAX_LEN` bytes.
-#[cfg(feature = "std")]
+#[cfg(any(feature = "std", not(no_core_net)))]
 macro_rules! serialize_display_bounded_length {
     ($value:expr, $max:expr, $serializer:expr) => {{
         let mut buffer = [0u8; $max];
-        let remaining_len = {
-            let mut remaining = &mut buffer[..];
-            write!(remaining, "{}", $value).unwrap();
-            remaining.len()
+        let written_len = {
+            let mut w = Wrapper {
+                buf: &mut buffer,
+                offset: 0,
+            };
+            write!(&mut w, "{}", $value).unwrap();
+            w.offset
         };
-        let written_len = buffer.len() - remaining_len;
         let written = &buffer[..written_len];
```

Comment:
This feels like something that could be written in a more elegant way?

Any inputs would be welcomed!

Reply:
Why was the wrapper struct needed at all? What std thing did the previous code use that core doesn't have?

Reply:
It results in 
```
error[E0599]: the method `write_fmt` exists for mutable reference `&mut [u8]`, but its trait bounds were not satisfied
   --> serde/src/ser/impls.rs:782:20
    |
782 |             write!(remaining, "{}", $value).unwrap();
    |                    ^^^^^^^^^
...
964 |             serialize_display_bounded_length!(self, MAX_LEN, serializer)
    |             ------------------------------------------------------------
    |             |
    |             method cannot be called on `&mut [u8]` due to unsatisfied trait bounds
    |             in this macro invocation
    |
note: must implement `io::Write`, `fmt::Write`, or have a `write_fmt` method
   --> serde/src/ser/impls.rs:782:20
    |
782 |             write!(remaining, "{}", $value).unwrap();
    |                    ^^^^^^^^^
...
964 |             serialize_display_bounded_length!(self, MAX_LEN, serializer)
    |             ------------------------------------------------------------ in this macro invocation
    = note: the following trait bounds were not satisfied:
            `[u8]: Write`
            which is required by `&mut [u8]: Write`
    = note: this error originates in the macro `serialize_display_bounded_length` (in Nightly builds, run with -Z macro-backtrace for more info)
    ```
    
Not sure if this is me doing something obviously wrong?

Reply:
There's a Write import guarded with cfg(std) somewhere. Generalize it to work with not(cfg(no_core_ip)), too

Reply:
I am not sure that there is? It seems that `core::fmt::Write` is imported as `FmtWrite`, But this seems like a complete copy of my wrapper (https://github.com/serde-rs/serde/blob/master/serde/src/de/format.rs), so I would assume it is indeed needed, but that I can reuse that one?

Reply:
Ah yea, that makes sense. Reusing existing abstractions makes sense